python-checks/src/test/java/org/sonar/python/checks/AbstractDuplicateKeyCheckTest.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.python.checks;
+
+import com.sonar.sslr.api.AstNode;
+import java.math.BigDecimal;
+import java.util.function.Supplier;
+import org.junit.Test;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.ExpressionStatement;
+import org.sonar.plugins.python.api.tree.FileInput;
+import org.sonar.python.parser.PythonParser;
+import org.sonar.python.tree.PythonTreeMaker;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.sonar.python.checks.AbstractDuplicateKeyCheck.NumericValue;
+import static org.sonar.python.checks.AbstractDuplicateKeyCheck.StringValue;
+
+@SuppressWarnings("squid:S2701")
+public class AbstractDuplicateKeyCheckTest {
+
+    private static <A> void testHashEquals(
+      Supplier<A> firstColliding,
+      Supplier<A> secondColliding,
+      Supplier<A> unrelated,
+      boolean canReturnDifferentValues
+    ) {
+        assertEquals(firstColliding.get().hashCode(), secondColliding.get().hashCode());
+        assertNotEquals(firstColliding.get(), secondColliding.get());
+        assertNotEquals(firstColliding.get(), unrelated.get());
+        if (!canReturnDifferentValues) {
+            assertEquals(firstColliding.get(), firstColliding.get());
+        } else {
+            assertNotEquals(firstColliding.get(), firstColliding.get());
+        }
+        A a = firstColliding.get();
+        assertEquals(a, a);
+        // This is not same as `assertNotNull`, it must invoke `equals`.
+        assertNotEquals(a, null);
```

Comment:
Maybe we could call `equals` directly and avoid the suppression in L38?

Reply:
Unfortunately, no.
If you use `assertFalse(...equals)`, it says to use `assertNotEquals` instead. Once you change it to `assertNotEquals`, it says to change it to `assertNotNull`. No matter what I did, the series of messages kept spinning until the test became trivial - it simply didn't believe me that the `equals` method is exactly what I want to test.

Reply:
Okay, that makes sense. Even though I still would be in favor of avoiding a class-level `SuppressWarnings` and marking the issue SQ raises as an FP.